runtime/src/system_instruction_processor.rs

```diff
@@ -483,6 +483,25 @@ pub fn process_instruction(
                 instruction_context.try_borrow_instruction_account(transaction_context, 0)?;
             authorize_nonce_account(&mut me, &nonce_authority, &signers, invoke_context)
         }
+        SystemInstruction::UpgradeNonceAccount => {
+            let separate_nonce_from_blockhash = invoke_context
+                .feature_set
+                .is_active(&feature_set::separate_nonce_from_blockhash::id());
+            if !separate_nonce_from_blockhash {
+                return Err(InstructionError::InvalidInstructionData);
+            }
+            instruction_context.check_number_of_instruction_accounts(1)?;
+            let mut nonce_account =
+                instruction_context.try_borrow_instruction_account(transaction_context, 0)?;
+            if nonce_account.get_owner() != &system_program::id() {
+                return Err(InstructionError::InvalidAccountOwner);
+            }
+            let nonce_versions: nonce::state::Versions = nonce_account.get_state()?;
+            match nonce_versions.upgrade() {
+                None => Err(InstructionError::InvalidArgument),
+                Some(nonce_versions) => nonce_account.set_state(&nonce_versions),
+            }
```

Comment:
:rotating_light:
The assumptions here are that:
1. Any extant system account that has non-zero data and is not a nonce account is invalid.
2. Writing to these invalid accounts is fine and won't break anything.

Reply:
Maybe I didn't see them but would be good to have a test to exercise these assumptions

Reply:
I believe this assumptions come from `get_system_account_kind`:
https://github.com/solana-labs/solana/blob/b4bdbb27d/runtime/src/system_instruction_processor.rs#L546-L563